ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -4139,27 +4141,42 @@ public List<Partition> getPartitions(Table tbl) throws HiveException {
   }
 
   /**
-   * Get all the partitions; unlike {@link #getPartitions(Table)}, does not include auth.
+   * Get all the partitions in batches; unlike {@link #getPartitions(Table)}, does not include auth.
    * @param tbl table for which partitions are needed
    * @return list of partition objects
    */
   public Set<Partition> getAllPartitionsOf(Table tbl) throws HiveException {
+    int batchSize= MetastoreConf.getIntVar(
+            Hive.get().getConf(), MetastoreConf.ConfVars.BATCH_RETRIEVE_MAX);
+    int maxRetries = HiveConf.getIntVar(conf, HiveConf.ConfVars.HIVE_GETPARTITIONS_MAX_RETRIES);
+    return getAllPartitionsOf(tbl, batchSize, 2, maxRetries);
+  }
+
+  public Set<Partition> getAllPartitionsOf(Table tbl, int batchSize, int decayingFactor,
+         int maxRetries) throws HiveException {
     if (!tbl.isPartitioned()) {
       return Sets.newHashSet(new Partition(tbl));
     }
-
-    List<org.apache.hadoop.hive.metastore.api.Partition> tParts;
+    Set<Partition> result = new LinkedHashSet<>();
+    RetryUtilities.ExponentiallyDecayingBatchWork batchTask = new RetryUtilities
+            .ExponentiallyDecayingBatchWork<Void>(batchSize, decayingFactor, maxRetries) {
+      @Override
+      public Void execute(int size) throws HiveException {
+        try {
+          result.clear();
+          new PartitionIterable(Hive.get(), tbl, null, size).forEach(result::add);
```

Comment:
We won't get partial or duplicate partitions due to some exception. In case of any exception, retry will happen by reducing the batch size. It will keep on retrying unless batchSize becomes 0 or maxRetry limit is exhausted. I have added test cases for all these 3 scenarios.